ocrd_network/ocrd_network/rabbitmq_utils/ocrd_messages.py

```diff
@@ -0,0 +1,117 @@
+            raise ValueError('input_file_grps must be set and contain at least 1 element')
+        if not (workspace_id or path_to_mets):
+            raise ValueError('Either `workspace_id` or `path_to_mets` must be set')
+
+        self.job_id = job_id  # uuid
+        self.processor_name = processor_name  # "ocrd-.*"
+        # Either of these two below
+        self.workspace_id = workspace_id  # uuid
+        self.path_to_mets = path_to_mets  # absolute path
+        self.input_file_grps = input_file_grps
+        self.output_file_grps = output_file_grps
+        # e.g., "PHYS_0005..PHYS_0010" will process only pages between 5-10
+        self.page_id = page_id if page_id else None
+        # e.g., "ocrd-cis-ocropy-binarize-result"
+        self.result_queue = result_queue_name if result_queue_name else (self.processor_name + "-result")
+        # processor parameters
+        self.parameters = parameters if parameters else None
+        self.created_time = created_time
+
+    @staticmethod
+    def encode_yml(ocrd_processing_message: OcrdProcessingMessage) -> bytes:
+        """convert OcrdProcessingMessage to yml
+        """
+        return yaml.dump(ocrd_processing_message.__dict__, indent=2).encode('utf-8')
+
+    @staticmethod
+    def decode_yml(ocrd_processing_message: bytes) -> OcrdProcessingMessage:
+        """Parse OcrdProcessingMessage from yml
+        """
+        msg = ocrd_processing_message.decode('utf-8')
+        data = yaml.load(msg, Loader=yaml.Loader)
+        return OcrdProcessingMessage(
+            job_id=data.get('job_id', None),
+            processor_name=data.get('processor_name', None),
+            created_time=data.get('created_time', None),
+            path_to_mets=data.get('path_to_mets', None),
+            workspace_id=data.get('workspace_id', None),
+            input_file_grps=data.get('input_file_grps', None),
+            output_file_grps=data.get('output_file_grps', None),
+            page_id=data.get('page_id', None),
+            parameters=data.get('parameters', None),
+            result_queue_name=data.get('result_queue', None),
+        )
+
+    @staticmethod
+    def from_job(job: Job) -> OcrdProcessingMessage:
+        return OcrdProcessingMessage(
+            job_id=job.id,
+            processor_name=job.processor_name,
+            path_to_mets=job.path,
+            input_file_grps=job.input_file_grps,
+            output_file_grps=job.output_file_grps,
+            page_id=job.page_id,
+            parameters=job.parameters,
+        )
+
+
+class OcrdResultMessage:
+    def __init__(self, job_id: str, status: str, workspace_id: Optional[str] = None,
+                 path_to_mets: Optional[str] = None) -> None:
+        self.job_id = job_id
+        self.status = status
+        # Either of these two below
+        self.workspace_id = workspace_id
+        self.path_to_mets = path_to_mets
+
+        if not (workspace_id or path_to_mets):
```

Comment:
Does it matter? No work is done with these variables. Anyway, I will transfer it before the assignments just to follow the same standard (the `OcrdProcessingMessage` also does the checks before the assignments).

Reply:
>> Exactly, but we do nothing with the data yet.

> We did.
> self.workspace_id = workspace_id
self.path_to_mets = path_to_mets

These two assignments do nothing to change the execution order (and potentially produce errors) neither has side effects.

> we ran some code which can be avoided if the validation failed.

Right! We are saving a few CPU cycles from the assignments when the validation fails. This happens almost never (unless unpredictable states) since the processing worker is the one creating the `OcrdResultMessage`, and hence, always assigning either of the fields. 

Anyway, I am not sure what is tried to be achieved with this debate here. The same amount of dedication for details is required in https://github.com/OCR-D/core/discussions/1015. In case the CPU cycles are important (IMO, they always are) there are a few other places where the code can be optimized to not run unnecessary code.

For example:
1. The input/output file groups (entry point to a NF workflow script) `must` be passed as comma-separated strings from the CLI
2. The Nextflow script has to convert these strings into lists of strings because that is what the `JobInput` validator expects according to the message scheme in the [spec](https://github.com/OCR-D/spec/blob/master/web_api.md). The reason they are lists: 
```For input-file-grp and output-file-grp, please keep them as lists. "Extra work" is just 1 line Python code to convert a list into a comma separated string :)``` [source](https://app.zenhub.com/workspaces/ocr-d-board-61b071467917f10021c2582a/issues/gh/ocr-d/zenhub/139). Where by "extra work" it was really meant to not run unnecessary code for every message.
3. Then these lists of strings are converted back again to a comma-separated string [here](https://github.com/OCR-D/core/pull/974/files/5ff8860a004a81f6e9fd9d5a93e51040bfc5766c#diff-57e0e8eb262bb1bd4331f5ee493caa5d412e804806d0e369739c73039a925523R290-R291) and [here](https://github.com/OCR-D/core/pull/974/files/5ff8860a004a81f6e9fd9d5a93e51040bfc5766c#diff-57e0e8eb262bb1bd4331f5ee493caa5d412e804806d0e369739c73039a925523R257-R258) to be able to pass them to the `ocrd_processor`.

The only benefit of having the input/output groups as lists is to make the spec feel more natural, but without any practical usability of them being lists.

Reply:
> Anyway, I am not sure what is tried to be achieved with this debate here.

What was tried to be achieved here is not about saving CPU cycle, but a good programming habit, writing good/maintainable code, starting from very small things.